### ISSUE TITLE
Correct MoE auxiliary-loss-free load balancing?

### ISSUE THREAD

**Issue by garrett361** [pytorch/torchtitan#1506]
A very small question: why is the second `expert_bias_delta` assignment used here?

https://github.com/pytorch/torchtitan/blob/cf30b2902718790cbe91900414c3201b6d7680b0/torchtitan/experiments/llama4/optimizer.py#L39-L43

This looks different than Algorithm 1 of https://arxiv.org/pdf/2408.15664, which would instead just be (IIUC):
```py
expert_bias_delta = moe.load_balance_coeff * torch.sign(
    moe.tokens_per_expert.mean() - moe.tokens_per_expert
)
moe.expert_bias.add_(expert_bias_delta)
```

CC @tianyu-l , who implemented this, I think.

**Comment by tianyu-l** [comment]
I think this update only keeps the mean of the `expert_bias` to be always 0, but wouldn't change token selection. Without this, the mean of `expert_bias` could shift to huge scale compared with the router raw scores themselves, which may impact training stability.

**Comment by garrett361** [comment]
Right, it doesn't change anything about what experts get chosen, since the expert bias is only used in the `top_k` computation, this line induces a scalar shift to the bias and `top_k` is invariant to such shifts:`top_k(x + scalar_bias) = top_k(x)`. 

https://github.com/pytorch/torchtitan/blob/cf30b2902718790cbe91900414c3201b6d7680b0/torchtitan/experiments/llama4/model/moe.py?plain=1#L172-L176

So I guess this line helps avoid `expert_bias` running off to +/- infinity and inducing `nan`s.
